### PR TITLE
Automatic DgsDataLoader naming

### PR DESCRIPTION
**Pull request by jord1e**
Pull request checklist
----

- [x] Please read our [contributor guide](https://github.com/Netflix/dgs-framework/blob/master/CONTRIBUTING.md)
- [x] Consider creating a discussion on the [discussion forum](https://github.com/Netflix/dgs-framework/discussions)
  first
- [x] Make sure the PR doesn't introduce backward compatibility issues
- [x] Make sure to have sufficient test cases

Pull Request type
----

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe):

Changes in this PR
----

This PR automatically names dataloaders, which is useful when using the `DgsDataFetchingEnvironment#getDataLoader(Class)` method.

It also implements automatic naming of field loaders, which is arguably less useful. If anyone objects to this we can error out on nameless field dataloaders.

Something to consider is that metric and tracing solutions may expose (field?)/class/package names of the dataloader. In the future we can add a `DgsDataLoaderNamingFactory` or something that gives library consumers the ability to modify the name programmatically (e.g. - only use the class name instead of the FQN) - this PR does not implement this.

Partially related to https://github.com/Netflix/dgs-framework/issues/861

Alternatives considered
----

Name the loaders

### CODE REVIEW

**Comment by srinivasankavitha** [comment]
Thanks for the PR. The changes look good to me, I need to test it out a bit before merging. Will do that this week.

**Comment by jord1e** [comment]
Please look at the automatic field naming the closest, I am seriously debating leaving it out (and throwing a exception on *unnamed* field dataloaders) because refactoring the field name breaks code (fields cannot be referenced as parameters in methods)

File: graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DgsDataLoaderProvider.kt [hunk 1]
```diff
@@ -37,10 +44,12 @@ import javax.annotation.PostConstruct
  */
 class DgsDataLoaderProvider(private val applicationContext: ApplicationContext) {
 
-    private val batchLoaders = mutableListOf<Pair<BatchLoader<*, *>, DgsDataLoader>>()
-    private val batchLoadersWithContext = mutableListOf<Pair<BatchLoaderWithContext<*, *>, DgsDataLoader>>()
-    private val mappedBatchLoaders = mutableListOf<Pair<MappedBatchLoader<*, *>, DgsDataLoader>>()
-    private val mappedBatchLoadersWithContext = mutableListOf<Pair<MappedBatchLoaderWithContext<*, *>, DgsDataLoader>>()
+    data class LoaderHolder<T>(val theLoader: T, val annotation: DgsDataLoader, val name: String)
```
Comment (jord1e) [inline graphql-dgs/src/main/kotlin/com/netflix/graphql/dgs/internal/DgsDataLoaderProvider.kt:47]: Should be private

**Comment by srinivasankavitha** [comment]
Regarding https://github.com/Netflix/dgs-framework/pull/904#issuecomment-1064340252

I agree, I think generating the names automatically for the field loaders feels like it may result in more confusion than help. Let's remove that and keep just the generation of class data loader names.

Otherwise looks good to me. Thanks for the PR!

**Review by srinivasankavitha — CHANGES_REQUESTED** [review]
We can merge the PR after you get a chance to remove the generation of names for field data loaders. Thanks!

**Comment by jord1e** [comment]
Made the following changes:

- Fields annotated with nameless @DgsDataLoader now throw `DgsUnnamedDataLoaderOnFieldException` upon initialization
- Added javadoc/ktdoc
- DgsDataLoaderProvider refactoring to DRY up some code (did I just invent a new verb?)
- Replaced the `prefix+_+fqnClassName` with just `Class.getSimpleName` (this is unambiguous enough in my opinion)
